Add Airbyte Protocol V1 support.

gosusnp opened the pull request:
## What

Enable Airbyte Protocol V1 in the platform.

Closes #16814 

## How

* Use the default objects as v1 and the v0 objects for v0 (as expected)
* Update the config to enable v0 and v1 in the platform

This leverages the fact that v1 doesn't change the structure of v0. However, we won't be fully using v1 until https://github.com/airbytehq/airbyte/pull/19240 is merged.
In order to properly use the v1 objects, we need to have a few tasks to decouple Java connectors from the default version of the protocol objects.

## Recommended reading order
Any

## 🚨 User Impact 🚨

Following this change, as expected but worth highlighting, we will be converting all messages v0->v1->v0 for all connectors until they have been migrated to v1.

gosusnp committed:
Add Airbyte Protocol V1 support.

jdpgrailsdev commented on airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/AirbyteMessageMigrationV1.java:
```diff
@@ -0,0 +1,38 @@
+
+  @Override
+  public Version getCurrentVersion() {
+    return new Version("1.0.0");
```
Does it make sense to define these versions somewhere in a constants class?  Something like PREVIOUS and CURRENT, as well as specific version numbers so that when we want to change, we can just basically update that class (e.g. point CURRENT to a newly added VERSION_X constant).  Feels like otherwise we are going to have a few versions of these strings floating around.  Not sure exactly how to reduce the duplication, but it feels like centralizing these would help.

gosusnp replied:
Those values are tied to migration itself. Currently, we can add V2 that would have a previous v1, current v2.

We currently have https://github.com/airbytehq/airbyte/blob/4974855b415ea4812e0d6647efb367ad837d8733/airbyte-commons/src/main/java/io/airbyte/commons/version/AirbyteProtocolVersion.java#L7, I could add some constants `V0`, `V1` in there.

gosusnp replied:
Updated

gosusnp committed:
Fix VersionedAirbyteStreamFactoryTest

gosusnp committed:
Remove AirbyteMessageMigrationV0 example

gosusnp committed:
Add Protocol Version constants

edgao commented on airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/AirbyteMessageMigrationV1.java:
```diff
@@ -0,0 +1,39 @@
+ * Placeholder AirbyteMessage Migration from v0 to v1
+ */
+@Singleton
+public class AirbyteMessageMigrationV1 implements AirbyteMessageMigration<AirbyteMessage, io.airbyte.protocol.models.AirbyteMessage> {
```
is it intentional to upgrade to `io.airbyte.protocol.models.AirbyteMessage` rather than `io.airbyte.protocol.models.v1.AirbyteMessage`?

gosusnp replied:
Yes, it is due to the fact that Java connectors and Platform share common libraries that are currently not versioning friendly.

In order to use `io.airbyte.protocol.models.v1.AirbyteMessage`, we need to update pretty much all the protocol objects used in the platform to `io.airbyte.protocol.models.v1.*` which forces us to currently bump some of the Java connectors to use `v1` objects as well. (see the slack thread from yesterday).

The idea here is to take a two step approach https://github.com/airbytehq/airbyte/issues/20037 and https://github.com/airbytehq/airbyte/issues/20038 rather than a giant change to bump platform to v1 and connectors to v0.

gosusnp replied:
Yes, the idea here is to have a two step change, enabling v1 on the default objects allows us to have a smaller PR focused on the actual changes to enable v1.
I will have another PR to switch the objects to v1.

benmoriceau reviewed (commented):
LGTM, I'll let previous review to approve

edgao committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way

edgao committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way

etsybaev committed:
🎉Updated normalization to handle new datatypes (#19721)

* Updated normalization simple stream processing to handle new datatypes

* Updated normalization nested stream processing to handle new datatypes

* Updated normalization nested stream processing to handle new datatypes

* Updated normalization drop_scd_catalog processing to handle new datatypes

* Updated normalization ephemeral test processing to handle new datatypes

* fixed more tests for normalization

* fixed more tests for normalization

* fixed more tests for normalization

* fixed more tests for normalization

* fixed more issues

* fixed more issues (clickhouse)

* fixed more issues

* fixed more issues

* fixed more issues

* added binary type processing for some DBs

* cleared commented code and moved some hardcodes to processing as macro

* fixed codestyle and cleared commented code

* minor refactor

* minor refactor

* minor refactor

* fixed bool cast error

* fixed dict->str cast error

* fixed is_combining_node cast py check

* removed commented code

* removed commented code

* committed autogenerated normalization_test_output files

* committed autogenerated normalization_test_output files (new files)

* refactored utils.py

* Updated utils.py to use Callable functions and get rid of property_type in is_number and is_bool functions

* committed autogenerated normalization_test_output files (new files)

* fixed typo in TIMESTAMP_WITH_TIMEZONE_TYPE

* updated stream_processor to handle string type first as a wider type

* fixed arrays normalization by updating is_simple_property method as per new approaches

* format

Co-authored-by: Edward Gao <edward.gao@airbyte.io>

gosusnp committed:
Update airbyte protocol migration (#20745)

* Extract MigrationContainer from AirbyteMessageMigrator

* Add ConfiguredAirbyteCatalogMigrations

* Add ConfiguredAirbyteCatalog to AirbyteMessageMigrations

* Enable ConfiguredAirbyteCatalog migration

* Fix tests

* Remove extra this.

* Add missing docs

* Typo

Co-authored-by: Edward Gao <edward.gao@airbyte.io>

grishick commented on airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/VersionedProtocolSerializer.java:
```diff
@@ -0,0 +1,32 @@
+
+  @Override
+  public String serialize(final ConfiguredAirbyteCatalog configuredAirbyteCatalog) {
+    return Jsons.serialize(configuredAirbyteCatalogMigrator.downgrade(configuredAirbyteCatalog, protocolVersion));
```
Shy is `serialize` always calling `downgrade` and what happens if `protocolVersion` the latest version?

gosusnp replied:
The Migrator figure out the migration path, if the version is current, nothing happens.

grishick replied:
Is there a scenario under which `serialize` will have to upgrade rather than downgrade a catalog? E.g. if the connector supports only the newer version? And if that's the scenario - does `downgrade` then essentially upgrade instead of `downgrading`?

grishick commented on airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/ConfiguredAirbyteCatalogMigrationV1.java:
```diff
@@ -0,0 +1,37 @@
+
+  @Override
+  public io.airbyte.protocol.models.ConfiguredAirbyteCatalog upgrade(ConfiguredAirbyteCatalog message) {
+    return Jsons.object(Jsons.jsonNode(message), io.airbyte.protocol.models.ConfiguredAirbyteCatalog.class);
```
trying to understand what's going on here, is `io.airbyte.protocol.models.ConfiguredAirbyteCatalog` identical to the latest `io.airbyte.protocol.models.v1.ConfiguredAirbyteCatalog`? Should anyone be using `io.airbyte.protocol.models.v1.ConfiguredAirbyteCatalog` directly outside of migration or should people always be using `io.airbyte.protocol.models.ConfiguredAirbyteCatalog`?

gosusnp replied:
This is currently a transition phase. Eventually, we should be using `io.airbyte.protocol.models.vX` and the objects without an explicit versions will be removed.

edgao committed:
Data types update: Implement protocol message migrations (#19240)

* Extract MigrationContainer from AirbyteMessageMigrator

* Add ConfiguredAirbyteCatalogMigrations

* Add ConfiguredAirbyteCatalog to AirbyteMessageMigrations

* Enable ConfiguredAirbyteCatalog migration

* set up scaffolding

* [wip] more scaffolding, basic unit test

* minimal green code

* [wip] add failing test for other primitive types

* correct version number

* handle basic primitive type decls

* add implicit cases

* add recursive schema

* formatting

* comment

* support not

* fix indentation

* handle all nested schema cases

* handle boolean schemas

* verify empty schema handling

* cleanup

* extract map

* code organization

* extract method

* reformat

* [wip] more tests, minor fix type array handling

* corrected test

* cleanup

* reformat

* switch to v1

* add support for multityped fields

* missed test case

* nested test class

* basic record upgrade

* implement record upgrades

* slight refactor

* comments+clarificationso

* extract constants

* (partly) correct model classes

* add de/ser

* formatting

* extract constants

* fix json reference

* update docs

* switch to v1 models

* fix compile+test

* add base64 handling

* use vnull

* Data types update: Implement protocol message downgrade path (#19909)

* rough skeleton for passing catalog into migration

* basic test

* more scaffolding

* basic implementation

* add primitives test

* add in other tests (nested fields currently failing)

* add formats

* impleent oneOf handling

* formatting

* oneOf handling

* better tests

* comments + organization

* progress

* basic test case

* downgrade objects, ish

* basic array implementation

* handle numeric failure

* test for new type

* handle array items

* empty schema handling

* first pass at oneof handling

* add more tests+handling

* more tests

* comments

* add empty oneof test case

* format + reorganize

* more reorganize

* fix name

* also downgrade binary data

* only import vnull

* move migrations into v1 package

* extract schema mutation code

* comment

* extract schema migration to new class

* extract record downgrade logic for future use

* format

* fix build after rebase

* rename private method for consistency

* also implement configuredcatalog migrations >.>

* quick and dirty tests

* slight cleanup

* fix tests

* pmd

* pmd test

* null check on message objects

* maybe fix acceptance tests?

* fix name

* extract constants

* more fixes

* tmp

* meh

* fix cdc acc tests

* revert to master source-postgres

* remove log messages

* revert other misc hacks

* integers are valid cursors

* remove unrelated change

* fix build

* fix build more?

* [MUST REVERT] use dev normalization

* capture kube logs

* also here?

* no debug logs?

* delete dup from merging

* add final everywhere

* revert test changes

Co-authored-by: Jimmy Ma <jimmy@airbyte.io>

gosusnp committed:
Merge remote-tracking branch 'origin/master' into gosusnp/platform-use-protocol-v1-the-quick-way

gosusnp committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way

gosusnp committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way

gosusnp committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way

gosusnp committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way

gosusnp committed:
On-the-fly migrations of persisted catalogs (#21757)

* On the fly catalog migration for normalization activity

* On the fly catalog migration for job persistence

* On the fly migration for standard sync persistence

* On the fly migration for airbyte catalogs

* Refactor code to share JsonSchema traversal

* Add V0 Data type search function

* PMD and Format

* Fix getOrInsertActorCatalog and ConfigRepositoryE2E tests

* Null-proofing CatalogMigrationV1Helper

* More null checks

* Fix test

* Format

* Add data type v1 support to the FE

* Changes AC test check to check exited ps (#21672)

some docker compose changes no longer show exited
processes.  this broke out test

this change should fix master

tested in a runner that failed

* Move wellknown types mapping to the utility function

* use protocolv1 normalization

---------

Co-authored-by: Topher Lubaway <asimplechris@gmail.com>
Co-authored-by: Edward Gao <edward.gao@airbyte.io>

gosusnp committed:
Update protocol support range (#21996)

edgao committed:
bump normalization version to 0.3.0

gosusnp committed:
Add version check on normalization (#22048)

* Add normalization min version check

* Add visible for testing

gosusnp committed:
Merge branch 'master' into gosusnp/platform-use-protocol-v1-the-quick-way
